openid-4-verifiable-credential-issuance-1_0.md

```diff
@@ -1152,14 +1152,16 @@ The path formed following the steps above MUST point to a JSON document complian
+* `authorization_servers`: OPTIONAL. A JSON array of strings, where each string is an identifier of the OAuth 2.0 Authorization Server (as defined in [@!RFC8414]) the Credential Issuer relies on for authorization. Since Credential Issuer may be using different authorization server depending on the Grant Type, the Wallet should determine which Authorization Server to use using `grant_types_supported` Authorization Server metadata parameter (If omitted, the default value is [`authorization_code`]). If this parameter is omitted, the entity providing the Credential Issuer is also acting as the AS, i.e., the Credential Issuer's identifier is used as the OAuth 2.0 Issuer value to obtain the Authorization Server metadata as per [@!RFC8414].
 * `credential_endpoint`: REQUIRED. URL of the Credential Issuer's Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components.
 * `batch_credential_endpoint`: OPTIONAL. URL of the Credential Issuer's Batch Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Batch Credential Endpoint.
 * `deferred_credential_endpoint`: OPTIONAL. URL of the Credential Issuer's Deferred Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Deferred Credential Endpoint.
 * `credential_response_encryption_alg_values_supported`: OPTIONAL. JSON array containing a list of the JWE [@!RFC7516] encryption algorithms (`alg` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].
 * `credential_response_encryption_enc_values_supported`: OPTIONAL. JSON array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].
 * `require_credential_response_encryption`: OPTIONAL. If `require_credential_response_encryption` is `true`, the Credential Issuer requires additional encryption on top of TLS for the Credential Response and expects encryption parameters to be present in the Credential Request and/or Batch Credential Request. If `require_credential_response_encryption` is `false` or omitted, the Credential Issuer indicates that no additional Credential Response encryption is required. If `require_credential_response_encryption` is `true`, `credential_response_encryption_alg_values_supported` MUST also be provided.
 
+Note: When using relying on multiple Authorization Servers for authorization, Credential Issuer MUST be able to identify from which Authorization Server each Access Tokens is coming.
```

Comment:
I don't know if this sentence is really necessary (it seems to overlap with the general "resource endpoints must check the access token is valid for the resource being access", but if it is this might be a better wording:

```suggestion
Note: When the Credential Issuer accepts access tokens from more than one Authorization Servers, the Credential Issuer MUST be able to identify which Authorization Server issued the Access Token.
```

Reply:
When talking to DanielF, Brian, you and others at OSW about this, this sentence seemed to be the key, but maybe I misunderstood.

Reply:
I do vaguely remember that conversation but I can't remember why anyone felt it important.

Maybe I'm missing something but I think it's also a bit weird because I don't think it's important that credential issuer determines which authorization server issued an access token, I think it's only important that it verify that it verifies that the token was issued by a server that trusts to issue a token for that particular resource etc. (Which is kind a subtle difference, but I think if both ASs issue JWT access tokens signed by the same key then it's not important for the resource server to identify which of the servers issued the access token because it doesn't matter, the RS can trust it because it's signed with the key it expects.)

Reply:
In the general case of RFC6750 etc there's no defined way for an RS (Credential Issuer here) to determine the AS that issued an AT.  So an RS that accepts ATs from more than one AS will have to do some nonstandard tricks that might not always work to figure out the issuer. That was the trust of the conversation as I recall.

Reply:
Thanks Brian. So are we essentially trying to say something like "Implementing a resource server that accepts tokens from multiple authorization servers makes introspecting the token hard and (as of the time of writing) pretty much the only sensible way is to only do this is with authorization servers that use https://datatracker.ietf.org/doc/html/rfc9068 or a proprietary AS/RS collusion that allows the RS to figure out what permissions are granted by the token"?

Reply:
I'm not sure I want to get into judgements about what's sensible or not :) But basically yes.

Reply:
I can live without the word "sensible" there :-)

I'm not sure the current sentence in this PR really gets that nuance across. I'm not sure the suggested sentence I wrote in my previous comment is great either.

Reply:
@jogu `access tokens` -> `Access Tokens`

Reply:
not exactly happy with the last few phrases, but how about this..?

```suggestion
Note: It can be hard for the Credential Issuer that accepts tokens from multiple Authorization Servers to introspect the Access Token to identify its issuer and permissions granted. Few ways to do so are either by relying on the Authorization Servers that use [@!RFC9068] or by Credential Issuer learning to understand proprietary Access Token structure of Authorization Server.
```

Reply:
Building upon Kristina's suggestion this might be slightly better:

```suggestion
Note: It can be challenging for a Credential Issuer that accepts tokens from multiple Authorization Servers to introspect an Access Token to check the validity and determine the permissions granted. Some ways to achieve this are relying on Authorization Servers that use [@!RFC9068] or by the Credential Issuer understanding the proprietary Access Token structures of the Authorization Servers.
```

(I've deliberately removed the mention of 'identify the issuer' - I do remember that being a key part of the OSW discussion, but I don't think we fully finished the discussion - the important aspect is that the RS is able to validate the token and get the permissions granted, and there are proprietary ways to do that if the ASes & RS are all run by the same party.)

Reply:
@jogu's text looks good to me